### ISSUE TITLE
Merging any tokens in a document corrupts the model and changes it dep output for other documents

### ISSUE THREAD

**Issue by mikkoairaksinen** [explosion/spaCy#7768]
## How to reproduce the behaviour
```
import unittest
import spacy


class TempTest(unittest.TestCase):
    string1 = "car comprising\nwheel,\nwindow"
    string2 = "This is another completely unrelated piece of text"

    nlp = spacy.load("en_core_web_lg")

    def test_retokenize_and_merge_corrupts_model(self):
        first_doc = self.nlp(self.string1)
        self.assertEqual("ROOT", first_doc[0].dep_)

        other_doc = self.nlp(self.string2)
        with other_doc.retokenize() as retokenizer:
            retokenizer.merge(other_doc[0:2])

        first_doc_after = self.nlp(self.string1)
        self.assertEqual("ROOT", first_doc_after[0].dep_)
```
We noticed this in an unit test which works when ran alone but fails when ran with other tests, and managed to pinpoint the problem to the fact that the spacy model object is shared between tests and running any test that merges any tokens in any document will corrupt that model and it will start outputting the wrong (or at least different) dep for the "car" token in the test string.

Might be the same bug as https://github.com/explosion/spaCy/issues/7662

## Your Environment
- **spaCy version:** 3.0.5
- **Platform:** Linux-5.8.0-48-generic-x86_64-with-glibc2.31
- **Python version:** 3.9.2
- **Pipelines:** en_core_web_lg (3.0.0)

**Comment by adrianeboyd** [comment]
Yes, this is the same bug as in #7662.

**Comment by adrianeboyd** [comment]
If you're training your own models, this fix will be in the upcoming v3.0.6 if you use the new default config settings.

Unfortunately there's really no way to fix existing pretrained pipelines because they've learned that this random/rare vector was UNK rather than a 0-vector and there's not currently a way to specify a particular row as the UNK vector, either.

If you need consistent predictions for your test suite, you can get them by switching to `spacy.MultiHashEmbed.v2` in your config for an existing pipeline (none of the config parameters have changed so editing this by hand won't break anything), but the predictions will probably be worse overall because of the different UNK behavior.

**Comment by mikkoairaksinen** [comment]
@adrianeboyd We are now looking into upgrading to spacy 3.1. We installed spacy 3.1 and the 3.1 models, but this test still fails. I thought that this bug was fixed in 3.1? Is the config change to `spacy.MultiHashEmbed.v2` still something that has to be done for tests to be consistent?
